### PR TITLE
Refactor ns1_zone module

### PR DESCRIPTION
**Pull request by mburtless**
Refactors diff logic, parameter sanitization and adds handling for subparams.  Also adds preliminary unit tests.

Note this is a _breaking_ change, as new/modified zone data is now returned in a context specific dict within Ansible results named `zone` instead of a generic dict named `data`.  Also should be considered breaking as this adds structured subparams for `secondary`, instead of accepting arbitrary keys.

### CODE REVIEW

File: library/ns1_zone.py [hunk 1]
```diff
@@ -7,8 +7,10 @@
 
 __metaclass__ = type
 
-
-from ansible.module_utils.ns1 import NS1ModuleBase, HAS_NS1
+try:
+    from ansible.module_utils.ns1 import NS1ModuleBase, HAS_NS1
+except ImportError:
+    from module_utils.ns1 import NS1ModuleBase, HAS_NS1 # noqa
```
Comment (rupa) [inline library/ns1_zone.py]: this is a 2/3 thing?

File: library/ns1_zone.py [hunk 2]
```diff
@@ -135,28 +168,26 @@
 
 
 try:
-    from ns1 import NS1, Config
     from ns1.rest.errors import ResourceException
 except ImportError:
     # This is handled in NS1 module_utils
     pass
```
Comment (rupa) [inline library/ns1_zone.py]: just wanna make sure this except is still needed - similar but different enough from the one above that i just wanna make sure it's still "load-bearing"

File: library/ns1_zone.py [hunk 3]
```diff
@@ -135,28 +168,26 @@
...
-    'secondary_enabled',
-    'primary_ip',
-    'primary_port',
-    'primary',
-    'primary_enabled',
-    'secondaries',
-    'dnssec'
+]
+
+# list of keys that should be ignored for calls to API
```
Comment (rupa) [inline library/ns1_zone.py]: can we make this comment more precise about what "ignored" means? So future spelunkers don't have to think about it. maybe `list of keys that will be removed from the data passed to API` or something

File: library/ns1_zone.py [hunk 4]
```diff
@@ -195,43 +245,61 @@ def __init__(self):
...
-        self.exec_module()
 
-    def api_params(self):
-        params = dict(
-            (key, self.module.params.get(key))
-            for key in ZONE_KEYS
-            if self.module.params.get(key) is not None
-        )
-        return params
+    def sanitize_params(self, params):
```
Comment (rupa) [inline library/ns1_zone.py]: we've all written dozens of these funcs. this is an improved name, but I really like calling them `prepare_for_api` - I think it's descriptive and has the good connotations from both the old name and your improved name.

Another option i don't mind, that arguably better when it's a consistent pattern is to make the method `serialize(self, deserialized)` and name the return value `serialized` - but that's not a great fit here since a good serializer should return a string. (But i like it as a pattern cause you just put any custom mangling in your serialize/deserialize methods - crispy like the way Go marshal/unmarshal works)

TL;DR - how does `prepare_for_api` hit you :)

File: library/ns1_zone.py [hunk 5]
```diff
@@ -243,39 +311,50 @@ def update(self, zone):
 
         self.module.exit_json(changed=changed, id=zone['id'], data=zone.data)
 
+    def create(self, zone):
+        if self.module.check_mode:
+            # short circuit in check mode
```
Comment (rupa) [inline library/ns1_zone.py]: wish we didn't need these comments but i can't think of an alternative that isn't clunkier in other ways. so, glad you put 'em there

Comment (mburtless) [reply library/ns1_zone.py]: Still refactoring program flow to get to the point where there is only one happy path through the module.  Think we can remove this once the module only exits in a single place.

File: library/ns1_zone.py (hunk 4 above)
Comment (mburtless) [reply library/ns1_zone.py]: I'm down with changing this to a more descriptive name, but take a look at [this](https://github.com/ns1/ns1-ansible-modules/blob/396abc160f4473572c9996d2f182ff84e9e711b0/library/ns1_zone.py#L298) to see another way this method is now being used (to normalize params for comparison against a retrieved zone).

Still like `prepare_for_api`, given that this is a more general normalization function now, and not just used right before submitting params to the API?

File: library/ns1_zone.py (hunk 2 above)
Comment (mburtless) [reply library/ns1_zone.py]: Yup, still needed.  The reason this `except` is here is because we want to throw a graceful failure via Ansible when the user is missing a required lib and not just throw a python error.

Since this code will be required in every module, we've made this DRY by also catching the [ImportError](https://github.com/ns1/ns1-ansible-modules/blob/396abc160f4473572c9996d2f182ff84e9e711b0/module_utils/ns1.py#L25) in our module_util and handling the failure there.

This is a pattern I stole from Ansible core modules, but I'm down to improve this if you have any idea?

Comment (rupa) [reply library/ns1_zone.py]: nope, just wanted to make sure it wasn't dangling and didn't spare the cycles to find out for myself :)

File: library/ns1_zone.py (hunk 4 above)
Comment (rupa) [reply library/ns1_zone.py]: good point, down with the name as you have it since it's not a serialization nor is it strictly for API mangling.

File: library/ns1_zone.py (hunk 1 above)
Comment (mburtless) [reply library/ns1_zone.py]: Ooh good catch, will add a comment.  The import path changes depending on if this is being executed via Ansible or directly via pytest.

File: module_utils/ns1.py [hunk 6]
```diff
@@ -1,40 +1,45 @@
 # -*- coding: utf-8 -*-
 
-# Copyright: (c) 2019, Matthew Burtless <mburtless@ns1.com>
+# Copyright: (c) 2020, NS1
```
Comment (rupa) [inline module_utils/ns1.py]: humility noted :)

File: module_utils/ns1.py [hunk 7]
```diff
@@ -1,40 +1,45 @@
...
 
 __metaclass__ = type
 
+import functools  # noqa
 from ansible.module_utils.basic import AnsibleModule
 
 try:
     from ansible.module_utils.basic import missing_required_lib
 except ImportError:
+
```
Comment (rupa) [inline module_utils/ns1.py]: did linter add this whitespace?

File: module_utils/ns1.py [hunk 8]
```diff
@@ -65,3 +70,25 @@ def errback(args):
...
+        :rtype: func
+        """
+
+        @functools.wraps(func)
+        def wrapper(self, *args, **kwargs):
+            if self.module.check_mode:
+                return
+            return func(self, *args, **kwargs)
+
+        return wrapper
```
Comment (rupa) [inline module_utils/ns1.py]: this is new but I don't see it being used? Also curious why you wrapped it in a class

Comment (mburtless) [reply module_utils/ns1.py]: This is used as a decorator on methods that make API calls in `ns1_zone.py` (i.e. https://github.com/ns1/ns1-ansible-modules/blob/ad0eb8a4e74610d847b5f933f7571e036bc3f432/library/ns1_zone.py#L324).  Can be reused when we refactor `ns1_record` and in new modules.

Shoved it in a class as a single place to keep our decorators in case we add more down the line (i.e. to make imports easier, etc).  Originally had it as class method of `NS1ModuleBase` but that opened a whole can of worms of issues I didn't want to deal with.

File: module_utils/ns1.py (hunk 7 above)
Comment (mburtless) [reply module_utils/ns1.py]: Yeah, this is a `black` thing.  Dumb.

File: library/ns1_zone.py [hunk 9]
```diff
@@ -110,173 +140,323 @@
 
 author:
   - 'Matthew Burtless (@mburtless)'
```
Comment (Zach-Johnson) [inline library/ns1_zone.py]: Not to rain on your parade but should this also be changed to NS1?

Comment (mburtless) [reply library/ns1_zone.py]: omg yes, good catch. Hold over from when we were trying to get this merged to core and had follow their module dev guide.

File: library/ns1_zone.py [hunk 10]
```diff
@@ -110,173 +140,323 @@
...
-            ):
-                changed = True
-                args[key] = self.module.params.get(key)
+                    zone = None
+        return zone
 
-        if self.module.check_mode:
-            # check mode short circuit before update
-            self.module.exit_json(changed=changed)
+    def compare_params(self, have, want):
```
Comment (Zach-Johnson) [inline library/ns1_zone.py]: should we put this somewhere else? I'm guessing this is going to be used by other resources once those are refactored? Or were you going to wait until we do that to pull this out somewhere common?

File: library/ns1_zone.py [hunk 11]
```diff
@@ -110,173 +140,323 @@
...
+
+        :param zone: Zone object of existing zone returned by NS1
+        :type zone: dict
+        :param want: Desired state of zone
+        :type want: dict
+        :return: Tuple in which first value reflects whether or not a change
+        occured and second value is new or updated zone object
+        :rtype: tuple(bool, dict)
+        """
+        changed = False
```
Comment (Zach-Johnson) [inline library/ns1_zone.py]: this `changed` stuff seems verbose to me, seems like this could just be 
```
if diff:
    return True, self.update(zone, diff)
return False, zone
```
But maybe that is not the Python way

File: library/ns1_zone.py (hunk 10 above)
Comment (mburtless) [reply library/ns1_zone.py]: You called it, was planning on waiting to pull this out until we refactor `ns1_record` module.

File: library/ns1_zone.py (hunk 9 above)
Comment (mburtless) [reply library/ns1_zone.py]: I'm actually not sure if `ansible-doc` will continue to work on this module if we don't match the format prescribed in https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-block.  Will test.

File: library/ns1_zone.py (hunk 11 above)
Comment (mburtless) [reply library/ns1_zone.py]: This is partly a holdover from the way this module was previously structured and partly a personal preference thing. My Python always tends to favor Single-Entry Single-Exit (as opposed to Go, where this would be tough to follow). It's def more verbose, but makes debugging easier imo.

In any case, I'm with you that it's probably unnecessary in a small funcs like these.

File: library/ns1_zone.py (hunk 9 above)
Comment (mburtless) [reply library/ns1_zone.py]: Looks like `ansible-doc` DGAF, but turned out I needed to remove the `doc_fragment` ref as well.  Both have been updated.

File: library/ns1_zone.py (hunk 11 above)
Comment (mburtless) [reply library/ns1_zone.py]: Added multiple returns
